src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs

```diff
@@ -205,26 +203,191 @@ bool tryBindAsHandlerType([NotNullWhen(true)] out BoundInterpolatedString? resul
             }
         }
 
+        private static bool AllInterpolatedStringPartsAreStrings(ImmutableArray<BoundExpression> parts)
+            => parts.All(p => p is BoundLiteral or BoundStringInsert { Value: { Type: { SpecialType: SpecialType.System_String } }, Alignment: null, Format: null });
+
+        private bool TryBindUnconvertedBinaryOperatorToDefaultInterpolatedStringHandler(BoundBinaryOperator binaryOperator, BindingDiagnosticBag diagnostics, [NotNullWhen(true)] out BoundBinaryOperator? convertedBinaryOperator)
+        {
+            // Much like BindUnconvertedInterpolatedStringToString above, we only want to use DefaultInterpolatedStringHandler if it's worth it. We therefore
+            // check for cases 1 and 2: if they are present, we let normal string binary operator binding machinery handle it. Otherwise, we take care of it ourselves.
+            Debug.Assert(binaryOperator.IsUnconvertedInterpolatedStringAddition);
+            convertedBinaryOperator = null;
+
+            var interpolatedStringHandlerType = Compilation.GetWellKnownType(WellKnownType.System_Runtime_CompilerServices_DefaultInterpolatedStringHandler);
+            if (interpolatedStringHandlerType.IsErrorType())
+            {
+                // Can't ever bind to the handler no matter what, so just let the default handling take care of it. Cases 4 and 5 are covered by this.
+                return false;
+            }
+
+            bool isConstant = true;
+            var stack = ArrayBuilder<BoundBinaryOperator>.GetInstance();
+            var partsArrayBuilder = ArrayBuilder<ImmutableArray<BoundExpression>>.GetInstance();
+            int partsCount = 0;
+
+            BoundBinaryOperator? current = binaryOperator;
+
+            while (current != null)
+            {
+                Debug.Assert(current.IsUnconvertedInterpolatedStringAddition);
+                stack.Push(current);
+                isConstant = isConstant && current.Right.ConstantValue is not null;
+                var rightInterpolatedString = (BoundUnconvertedInterpolatedString)current.Right;
+
+                if (rightInterpolatedString.Parts.ContainsAwaitExpression())
+                {
+                    // Exception to case 3. Delegate to standard binding.
+                    stack.Free();
+                    partsArrayBuilder.Free();
+                    return false;
+                }
+
+                partsCount += rightInterpolatedString.Parts.Length;
+                partsArrayBuilder.Add(rightInterpolatedString.Parts);
+
```

Comment:
Consider extracting this logic into a local function to be reused below.

Reply:
There's a bunch of methods that visit the binary ops to get to the interpolated strings. Maybe a visitor helper method would make sense? `BoundBinaryOperator.VisitUnconvertedInterpolatedStrings(...)`?

Reply:
I don't really see any particular logic that can be easily shared between the various visits. This location, for example, wants to collect the parts then rebind the interpolated strings, bailing out if it encounters bad data. The abstract flow pass needs to visit every piece. Local rewriting needs to collect the pieces then replace the original operator with a single BoundSequence. IMO, they're all too different to try and make a single helper for the logic.